soroban-env-host/src/native_contract/token/allowance.rs

```diff
@@ -37,11 +37,7 @@ pub fn write_allowance(
     // validates the expiration and then returns the ledger seq
     // The expiration can be less than ledger seq if clearing an allowance
     let ledger_seq = e.with_ledger_info(|li| {
-        if expiration
-            > li.sequence_number
-                .saturating_sub(1)
-                .saturating_add(li.max_entry_expiration)
-        {
+        if expiration > e.max_expiration_ledger()? {
```

Comment:
We generally gracefully limit too-large bumps to max expiration. However, in cases like this one we actually need to make sure that the entry can live long enough. I don't think this logic should be elsewhere, at least for the sake of more clear error reporting (IIUC the proposal is to enforce the limit when we try to flush the entries?)

Maybe it's worth to not allow bumping the temp entries past max expiration ledger though, to make sure it's not possible to have a bug where expect temp entry live time is no allowed. I think we can still gracefully handle persistent entry bumps though, as there is really no vulnerability here. Any thoughts @sisuresh ?

Reply:
Yeah I would think all entires have the same relationship between their expiry time and the max expiry time as one another, so this could/should be centralized when writing an entry. But maybe I'm misunderstanding, what's different about an allowance?

Reply:
Yeah this was done here for better error reporting, but I just realized that InvokeHostFunctionOp doesn't error if you try to exceed the max. Instead it caps the bump. But this is inconsistent from BumpFootprintOp, which will error if you bump past the max.

Reply:
I'm not opposed to returning an error for too large bumps (maybe limited to temp entries though). But we should return the error from the `bump` operation (that is not used here). I'm not sure we need additional enforcement from the core side (maybe just a sanity check?) - I wouldn't want to return an error that can be traced quite precisely only after transaction has almost finished. We can add this as a followup. This code (and nonce code) still should stay as is for the sake of error clarity.